File: counting_sort/cpp/counting_sort.cpp
/* Created by SubCoder1 on 31-May-18.

     Sorting Technique -- Counting Sort
     Input             -- A sequence of n numbers
     Output            -- Reordered into a definite sequence {Ascending(HERE)}

     Time Complexity :-
        Worst Case - O(N^2)
        Best  Case - O(N + RANGE)
 */





#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

 using namespace std::chrono;
 using namespace std;

int main(){
    int inpt;
    vector<int> store;
    vector<int>:: iterator it;

    cout << "Unsorted Input -- ";
    while(cin.peek() != '\n'){
        cin >> inpt;
        store.push_back(inpt);
    }
    
    auto start = high_resolution_clock::now();

    it = max_element(store.begin(), store.end());
    int range = *it;
    vector<int> aux(range, 0);

    for(auto i: store){ aux[i-1]+=1; }

    cout << "Sorted Output -- ";
    for(int i = 0; i < range; i++){
        for(int j = 1; j <= aux[i]; j++){
            cout << i+1 << " ";
        }
    }

    auto end = high_resolution_clock::now();

    cout << "Time Elapsed -- " << duration_cast<microseconds>(end - start).count() << " microS.\n";

}

File: counting_sort/python/counting_sort.py
"""   Sorting Technique -- Counting Sort
      Input             -- A range(Max element in the seq. of array + 1) and a sequence of n numbers
      Output            -- Reordered into a definite sequence {Ascending(HERE)}

      Time Complexity :-
         Worst Case - O(N^2)
         Best  Case - O(N + RANGE)   """



print("Enter Range -- ",end='')
Range = int(input())
arr = list(map(int, input().split()))
buckets = [0]*Range
sortedIndex = 0

for i in arr :
    buckets[i] += 1
for index in range(len(buckets)) :
    while buckets[index] > 0 :
        arr[sortedIndex] = index
        sortedIndex += 1
        buckets[index] -= 1
print("Sorted Output -- ", end='')
print(' '.join([str(x) for x in arr]))
